File: Java/src/_0543_Diameter_of_Binary_Tree/Solution.java
package _0543_Diameter_of_Binary_Tree;

/**
 * Date: 2021/7/16 21:21
 * Content:
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：38.3 MB, 在所有 Java 提交中击败了66.95%的用户
 */
public class Solution {

    int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return ans;
    }

    private int dfs(TreeNode root) {
        if (root == null) return 0;
        int left = dfs(root.left), right = dfs(root.right);
        ans = Math.max(ans, left + right);
        return Math.max(left, right) + 1;
    }

    public static void main(String[] args) {

        int nu = Integer.MIN_VALUE;  // 用Integer.MIN_VALUE表示二叉树 null
        int[] nums = {
                1,
                2, 3,
                4, 5
        };
        TreeNode root = (new MyTree(nums)).getRoot();
        System.out.println((new Solution()).diameterOfBinaryTree(root));  // 3
    }
}

File: Python/_0543_Diameter_of_Binary_Tree/Solution.py
# coding=utf-8
# Date: 2021/7/16 21:26
from MyTree import *


# 执行用时：48 ms, 在所有 Python3 提交中击败了84.54%的用户
# 内存消耗：16.7 MB, 在所有 Python3 提交中击败了81.80%的用户
class Solution:
    def __init__(self):
        self.ans = 0

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.dfs(root)
        return self.ans

    def dfs(self, root):
        if root is None:
            return 0
        l, r = self.dfs(root.left), self.dfs(root.right)
        self.ans = max(self.ans, l + r)
        return max(l, r) + 1


if __name__ == "__main__":
    nu = None
    nums = [
        1,
        2, 3,
        4, 5
    ]
    root = MyTree(nums).getRoot()
    print(Solution().diameterOfBinaryTree(root))  # 3
